# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/debugger/Frame.cpp
# Commit: d70ec1e79887
# Full Hash: d70ec1e7988710de0b79351f730a2498972cd7e0
# Author: Logan Smyth <loganfsmyth@gmail.com>
# Date: 2019-12-09 09:50:39
# Regressor Bug: 1600204
# File Overlap Count: 1
# Description:
#   Bug 1600204 - Part 1: Support .type on suspended generator frames. r=jimb
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D55231
# ==============================================================================

diff -r e3354d109039 -r d70ec1e79887 js/src/debugger/Frame.cpp
--- a/js/src/debugger/Frame.cpp	Fri Dec 06 21:44:23 2019 +0000
+++ b/js/src/debugger/Frame.cpp	Fri Dec 06 02:29:18 2019 +0000
@@ -665,30 +665,36 @@
 
 /* static */
 DebuggerFrameType DebuggerFrame::getType(HandleDebuggerFrame frame) {
-  AbstractFramePtr referent = DebuggerFrame::getReferent(frame);
+  if (frame->isOnStack()) {
+    AbstractFramePtr referent = DebuggerFrame::getReferent(frame);
 
-  // Indirect eval frames are both isGlobalFrame() and isEvalFrame(), so the
-  // order of checks here is significant.
-  if (referent.isEvalFrame()) {
-    return DebuggerFrameType::Eval;
-  }
+    // Indirect eval frames are both isGlobalFrame() and isEvalFrame(), so the
+    // order of checks here is significant.
+    if (referent.isEvalFrame()) {
+      return DebuggerFrameType::Eval;
+    }
+
+    if (referent.isGlobalFrame()) {
+      return DebuggerFrameType::Global;
+    }
 
-  if (referent.isGlobalFrame()) {
-    return DebuggerFrameType::Global;
-  }
+    if (referent.isFunctionFrame()) {
+      return DebuggerFrameType::Call;
+    }
 
-  if (referent.isFunctionFrame()) {
+    if (referent.isModuleFrame()) {
+      return DebuggerFrameType::Module;
+    }
+
+    if (referent.isWasmDebugFrame()) {
+      return DebuggerFrameType::WasmCall;
+    }
+  } else {
+    MOZ_ASSERT(frame->hasGenerator());
+
     return DebuggerFrameType::Call;
   }
 
-  if (referent.isModuleFrame()) {
-    return DebuggerFrameType::Module;
-  }
-
-  if (referent.isWasmDebugFrame()) {
-    return DebuggerFrameType::WasmCall;
-  }
-
   MOZ_CRASH("Unknown frame type");
 }
 
@@ -1307,7 +1313,7 @@
 }
 
 bool DebuggerFrame::CallData::typeGetter() {
-  if (!ensureOnStack()) {
+  if (!ensureOnStackOrSuspended()) {
     return false;
   }
 